lib/ts/BaseLogFile.cc

```diff
@@ -0,0 +1,571 @@
+  snprintf(roll_name, LOGFILE_ROLL_MAXPATHLEN, "%s%s%s.%s-%s%s", m_name, (m_hostname ? LOGFILE_SEPARATOR_STRING : ""),
+           (m_hostname ? m_hostname : ""), start_time_ext, end_time_ext, LOGFILE_ROLLED_EXTENSION);
+
+  // It may be possible that the file we want to roll into already
+  // exists.  If so, then we need to add a version tag to the rolled
+  // filename as well so that we don't clobber existing files.
+  int version = 1;
+  while (BaseLogFile::exists(roll_name)) {
+    log_log_trace("The rolled file %s already exists; adding version "
+                  "tag %d to avoid clobbering the existing file.\n",
+                  roll_name, version);
+    snprintf(roll_name, LOGFILE_ROLL_MAXPATHLEN, "%s%s%s.%s-%s.%d%s", m_name, (m_hostname ? LOGFILE_SEPARATOR_STRING : ""),
+             (m_hostname ? m_hostname : ""), start_time_ext, end_time_ext, version, LOGFILE_ROLLED_EXTENSION);
+    version++;
+  }
+
+  // It's now safe to rename the file.
+  if (::rename(m_name, roll_name) < 0) {
+    log_log_error("Traffic Server could not rename logfile %s to %s, error %d: "
+                  "%s.\n",
+                  m_name, roll_name, errno, strerror(errno));
+    return 0;
+  }
+
+  // reset m_start_time
+  m_start_time = 0;
+  m_bytes_written = 0;
+
+  log_log_trace("The logfile %s was rolled to %s.\n", m_name, roll_name);
+
+  return 1;
+}
+
+/*
+ * The more convienent rolling function. Intended use is for less
+ * critical logs such as diags.log or traffic.out, since _exact_
+ * timestamps may be less important
+ *
+ * The function calls roll(long,long) with these parameters:
+ * Start time is either 0 or creation time stored in the metafile,
+ * whichever is greater
+ *
+ * End time is the current time
+ *
+ * Returns 1 on success, 0 otherwise
+ */
+int
+BaseLogFile::roll()
+{
+  long start;
+  time_t now = time(NULL);
+
+  if (!m_meta_info || !m_meta_info->get_creation_time(&start))
+    start = 0L;
+
+  return roll(start, now);
+}
+
+
+/*
+ * This function will return true if the given filename corresponds to a
+ * rolled logfile.  We make this determination based on the file extension.
+ */
+bool
+BaseLogFile::rolled_logfile(char *path)
+{
+  const int target_len = (int)strlen(LOGFILE_ROLLED_EXTENSION);
+  int len = (int)strlen(path);
+  if (len > target_len) {
+    char *str = &path[len - target_len];
+    if (!strcmp(str, LOGFILE_ROLLED_EXTENSION)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/*
+ * Returns if the provided file in 'pathname' exists or not
+ */
+bool
+BaseLogFile::exists(const char *pathname)
+{
+  ink_assert(pathname != NULL);
+  return (pathname && ::access(pathname, F_OK) == 0);
+}
+
+/*
+ * Opens the BaseLogFile and associated BaseMetaInfo on disk if it exists
+ * Returns relevant exit status
+ */
+int
+BaseLogFile::open_file(int perm)
+{
+  log_log_trace("BaseLogFile: entered open_file()\n");
+  if (is_open()) {
+    return LOG_FILE_NO_ERROR;
+  }
+
+  if (m_name && !strcmp(m_name, "stdout")) {
```

Comment:
You assume later in this method that `m_name` is not `NULL`. So you should either assume that here as well, or do a pre-check verifying it is not `NULL` and failing out of the method if it is.